keps/sig-network/3070-reserved-service-ip-range/README.md

```diff
@@ -0,0 +1,751 @@
+	Data []byte
+}
+```
+
+The [apiserver Service registry contains allocators backed by those bitmaps](p[kg/registry/core/rest/storage_core.go](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/pkg/registry/core/rest/storage_core.go#L96-L103)) in order to assign these IPs.
+
+```go
+type LegacyRESTStorage struct {
+	ServiceClusterIPAllocator          rangeallocation.RangeRegistry
+	SecondaryServiceClusterIPAllocator rangeallocation.RangeRegistry
+	ServiceNodePortAllocator           rangeallocation.RangeRegistry
+}
+
+...
+func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generic.RESTOptionsGetter) (LegacyRESTStorage, genericapiserver.APIGroupInfo, error) {
+  ...
+	serviceClusterIPAllocator, err := ipallocator.New(&serviceClusterIPRange, func(max int, rangeSpec string) (allocator.Interface, error) {
+		mem := allocator.NewAllocationMap(max, rangeSpec)
+		// TODO etcdallocator package to return a storage interface via the storageFactory
+		etcd, err := serviceallocator.NewEtcd(mem, "/ranges/serviceips", serviceStorageConfig.ForResource(api.Resource("serviceipallocations")))
+		if err != nil {
+			return nil, err
+		}
+		serviceClusterIPRegistry = etcd
+		return etcd, nil
+	})
+```
+
+The bitmaps implement a [strategy for allocating a random free value if none is specified](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/pkg/registry/core/service/allocator/bitmap.go#L186-L205):
+
+```go
+// randomScanStrategy chooses a random address from the provided big.Int, and then
+// scans forward looking for the next available address (it will wrap the range if
+// necessary).
+type randomScanStrategy struct {
+	rand *rand.Rand
+}
+
+func (rss randomScanStrategy) AllocateBit(allocated *big.Int, max, count int) (int, bool) {
+	if count >= max {
+		return 0, false
+	}
+	offset := rss.rand.Intn(max)
+	for i := 0; i < max; i++ {
+		at := (offset + i) % max
+		if allocated.Bit(at) == 0 {
+			return at, true
+		}
+	}
+	return 0, false
+}
+```
+
+The proposal is to implement a new strategy on the Service IP allocation bitmap that avoids to use the first 32 values
+if there are free values on the upper block of the range.
+
+```go
+// randomScanReservedStrategy choose a random address from the provided big.Int and then scans
+// forward looking for the next available address. The big.Int range is subdivided so it will try
+// to allocate first from the reserved block of addresses (it will wrap the reserved subrange if necessary).
+// If there is no free address it will try to allocate one from the reserved block too.
+type randomScanReservedStrategy struct {
+	rand     *rand.Rand
+	reserved int
+}
+```
+
+In order to avoid any possible issue with cluster with small subnets, this strategy will only work if the service IP range
+is greater than 64, i.e., these apply only for Service subnets greater than IPv4 /26 or IPv6 /122.
```

Comment:
If your service CIDR only has 64 addresses in it, then the odds of someone randomly getting assigned the .10 address are much higher than in the case where you have 65,536 addresses, so it seems like the reserved range is even more necessary...

Reply:
64 is just is the next range available after 32, that is the number that we choose for the reserved range, so you have at least the same for static and reserved
We can choose different numbers, it just seems that 32 should be enough

Reply:
No, I mean, the basic premise of this KEP is "it's important for us to avoid the 1-in-65,536 chance of a Service getting assigned the IP that CoreDNS wants in a cluster with a /16 service CIDR" so it seems weird to say that we aren't going to solve the problem in the /26 case where it is 1,000 times more likely to happen...

Reply:
yeah, it seems more important for the smaller ranges